[Cocoa] Add a FairPlay keyserver and EME tests to LayoutTests

jernoble opened the pull request:
#### 809ad82ca6e22e19626ef696f3828ed39560bfc1
<pre>
[Cocoa] Add a FairPlay keyserver and EME tests to LayoutTests
<a href="https://bugs.webkit.org/show_bug.cgi?id=237731">https://bugs.webkit.org/show_bug.cgi?id=237731</a>
&lt;rdar://problem/90453705&gt;

Reviewed by Eric Carlson.

LayoutTests:
Add an implementation of a FairPlay keyserver, which can run as a CGI script as part of
the normal run-webkit-httpd, to LayoutTests.

The Encrypted Media Extensions API has traditionally required access to an external
keyserver to respond to requests by the client for the encryption keys needed to
decrypt media content. However, our testing infrastructure requires that no external
network access is allowed (aside from localhost) from the process being tested.

A keyserver which can run locally will enable WebKit developers to write and run layout
tests exercising both the EME API and platform encrypted media playback capabilities,
and allows those tests to be run on our bots as both pre-submission checks and
as regression tests.

The initial keyserver implementation is written in python3, has minimal package requirements,
and currently only supports lease-style liceneses. It can be extended to support rental-style
and persistable licenses. It currently only supports &quot;modern&quot; EME key requests, but can
be extended to support &quot;legacy&quot; EME key requests as well.

The initial set of tests validate successful key exchange with &apos;skd&apos;, &apos;cenc&apos;, and &apos;sinf&apos;
init data formats, key renewal, key rotation, multiple key support, and HLS playback.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-1.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-2.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-3.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-4.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-payload.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-1.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-2.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-3.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-4.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-payload.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/main.ts: Added.
* LayoutTests/http/tests/media/fairplay/content/prog_index.m3u8: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/prog_index.m3u8: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/segment-0.ts: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/segment-1.ts: Added.
* LayoutTests/http/tests/media/fairplay/eme2016.js: Added.
(async fetchBuffer):
(async getResponse):
(async startEME):
(async fetchAndAppend):
(async runAndWaitForLicenseRequest):
(async fetchAndWaitForLicenseRequest):
(async fetchAppendAndWaitForEncrypted):
(async createBufferAndAppend):
(async createBufferAppendAndWaitForEncrypted):
* LayoutTests/http/tests/media/fairplay/fps-hls-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-key-rotation-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-key-rotation.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-update-reject-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-update-reject.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-cenc-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-cenc.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-sinf-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-sinf.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-skd-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-skd.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-renewal-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-renewal.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-rotation-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-rotation.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-multiple-keys-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-multiple-keys.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-same-key-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-same-key.html: Added.
* LayoutTests/http/tests/media/fairplay/mse.js: Added.
(MSE):
(formatRangeHeader):
(MSE.prototype.start):
(MSE.prototype.handleSourceOpen):
(MSE.prototype.fetchNext):
(MSE.prototype.fetchChunk):
(MSE.prototype.handleChunkLoaded):
(MSE.prototype.handleChunkError):
(MSE.prototype.handleSourceBufferUpdateStart):
(MSE.prototype.handleSourceBufferUpdateEnd):
(MSE.prototype.removeAllSamples):
* LayoutTests/http/tests/media/fairplay/resources/GenerateCKC.py: Added.
* LayoutTests/http/tests/media/fairplay/resources/cert.der: Added.
* LayoutTests/http/tests/media/fairplay/resources/index.py: Added.
* LayoutTests/http/tests/media/fairplay/resources/keyserver/CKCContainer.py: Added.
(CKCContainer):
(CKCContainer.__init__):
(CKCContainer.version):
(CKCContainer.iv):
(CKCContainer.tllvs):
(CKCContainer.encryptedPayload):
(CKCContainer.addTLLV):
(CKCContainer.getTLLV):
(CKCContainer.encrypt):
(CKCContainer.serialize):
(CKCContainer.__str__):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Constants.py: Added.
(SKDServerAESEncType):
(SKDServerAESEncMode):
(SKDHDCPInformation):
(SKDClientRefTimePlayback):
(SKDLeaseRentalType):
(SKDStreamingIndicatorValue):
(SKDCapabilityFlags):
(Constants):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/DFunction.py: Added.
(ComputeDFunctionHash):
(Encrypt):
(DFunction):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/SPCContainer.py: Added.
(SPCContainer):
(SPCContainer.parse):
(SPCContainer.__init__):
(SPCContainer.spcVersion):
(SPCContainer.spcIV):
(SPCContainer.wrappedKey):
(SPCContainer.certificateHash):
(SPCContainer.payloadSize):
(SPCContainer.payloadData):
(SPCContainer.tllvs):
(SPCContainer.decrypt):
(SPCContainer.getTLLV):
(SPCContainer.checkIntegrity):
(SPCContainer.generateCKC):
(SPCContainer.generateARSeed):
(SPCContainer.__str__):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Structs.py: Added.
(Offset):
(SKDAssetInfo):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLV.py: Added.
(TLLV):
(TLLV.parse):
(TLLV.paddingForValueData):
(TLLV.__init__):
(TLLV.tagValue):
(TLLV.totalSize):
(TLLV.valueSize):
(TLLV.paddingSize):
(TLLV.valueData):
(TLLV.paddingData):
(TLLV.parseValue):
(TLLV.serialize):
(TLLV.__str__):
(Unknown):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py: Added.
(SK_R1):
(SK_R1.parseValue):
(SK_R1.iv):
(SK_R1.payload):
(SK_R1.isDecrypted):
(SK_R1.sk):
(SK_R1.hu):
(SK_R1.r1):
(SK_R1.integrity):
(SK_R1.decrypt):
(SK_R1.encrypt):
(SK_R1.__str__):
(SK_R1_Integrity):
(SK_R1_Integrity.parseValue):
(SK_R1_Integrity.value):
(SK_R1_Integrity.__str__):
(Capbilities):
(Capbilities.parseValue):
(Capbilities.capabilityFlags):
(Capbilities.has):
(Capbilities.set):
(Capbilities.clear):
(AntiReplaySeed):
(AntiReplaySeed.parseValue):
(AntiReplaySeed.value):
(R2):
(R2.parseValue):
(R2.value):
(ReturnTagRequest):
(ReturnTagRequest.parseValue):
(ReturnTagRequest.returnTags):
(AssetID):
(AssetID.parseValue):
(AssetID.value):
(TransactionID):
(TransactionID.parseValue):
(ProtocolVersionsSupported):
(ProtocolVersionsSupported.parseValue):
(ProtocolVersionsSupported.supportedVersions):
(ProtocolVersionUsed):
(ProtocolVersionUsed.parseValue):
(ProtocolVersionUsed.usedVersion):
(StreamingIndicator):
(StreamingIndicator.parseValue):
(StreamingIndicator.value):
(MediaPlaybackState):
(MediaPlaybackState.parseValue):
(MediaPlaybackState.creationDate):
(MediaPlaybackState.playbackState):
(MediaPlaybackState.sessionID):
(EncryptedContentKey):
(EncryptedContentKey.create):
(EncryptedContentKey.parseValue):
(EncryptedContentKey.iv):
(EncryptedContentKey.ck):
(EncryptedContentKey.__str__):
(R1):
(R1.create):
(R1.parseValue):
(R1.value):
(ContentKeyDuration):
(ContentKeyDuration.create):
(ContentKeyDuration.parseValue):
(ContentKeyDuration.leaseDuration):
(ContentKeyDuration.rentalDuration):
(ContentKeyDuration.keyType):
(ContentKeyDuration.__str__):
(HDCPEnforcement):
(HDCPEnforcement.parseValue):
(HDCPEnforcement.value):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py: Added.
(Utils):
(Utils.LoadPrivateKey):
(Utils.GenRandom):
(Utils.RSADecryptKey):
(Utils.AESEncryptDecrypt):
(Utils.FetchContentKeyAndIV):
(Utils.GetSupportedVersions):
(Utils.GetBigEndian32):
(Utils.GetBigEndian64):
(Utils.SetBigEndian32):
(Utils.SetBigEndian64):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/__init__.py: Added.
(cd_dot_dot):
(_maybe_add_scripts_path):
(_force_cryptography_install):
(_force_openssl_install):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_DFunction.py: Added.
(TestDFunction):
(TestDFunction.test_DFunctionHash):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_TLLV.py: Added.
(TestTLLV):
(TestTLLV.test_TLLV):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_TLLVs.py: Added.
(TestTLLVs):
(TestTLLVs.test_SK_R1):
(TestTLLVs.test_EncryptedContentKey):
(TestTLLVs.test_Capabilities):
(TestTLLVs.test_ReturnTagRequest):
(TestTLLVs.test_AssetID):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_Utils.py: Added.
(TestUtils):
* LayoutTests/http/tests/media/fairplay/support.js: Added.
(uInt8ArrayToString):
(base64DecodeUint8Array):
(stringToUInt8Array):
(base64EncodeUint8Array):
(failTestWithException):
* LayoutTests/platform/ios-simulator/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/253230@main">https://commits.webkit.org/253230@main</a>
</pre>

youssefsoli commented on LayoutTests/http/tests/media/fairplay/eme2016.js:
```diff
@@ -0,0 +1,105 @@
+    consoleWrite('PROMISE: licenseResponse resolved');
+    var license = await licenseResponse.text();
+
+    responseObject = JSON.parse(license.trim());
```
Either `response.json()` can be used instead of the `.text` `.trim` and `JSON.parse`
OR
based off the test failure and if it's feasible that the server fails, it might make sense to try/catch the JSON parse and dump the text if it fails, rather than having the cryptic `SyntaxError: JSON Parse error: Unrecognized token` error.

jernoble replied:
Okay, I'll add a try/catch block around this, and just re-use this method in all the other .html tests (rather than make the same change in each place).

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-hls-key-rotation.html:
```diff
@@ -0,0 +1,92 @@
+                });
+                var license = await licenseResponse.text();
+
+                responseObject = JSON.parse(license.trim());
```
ditto

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-hls-update-reject.html:
```diff
@@ -0,0 +1,79 @@
+        var license = await licenseResponse.text();
+        consoleWrite('PROMISE: licenseResponse resolved');
+
+        responseObject = JSON.parse(license.trim());
```
ditto

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-hls.html:
```diff
@@ -0,0 +1,84 @@
+        consoleWrite('PROMISE: licenseResponse resolved');
+        var license = await licenseResponse.text();
+
+        responseObject = JSON.parse(license.trim());
```
ditto

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-init-data-cenc.html:
```diff
@@ -0,0 +1,77 @@
+                }),
+            });
+            consoleWrite('PROMISE: licenseResponse resolved');
+            var license = await licenseResponse.text();
+
+            responseObject = JSON.parse(license.trim());
```
ditto

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-init-data-sinf.html:
```diff
@@ -0,0 +1,71 @@
+        consoleWrite('PROMISE: licenseResponse resolved');
+        var license = await licenseResponse.text();
+
+        responseObject = JSON.parse(license.trim());
```
ditto

youssefsoli commented on LayoutTests/http/tests/media/fairplay/fps-init-data-skd.html:
```diff
@@ -0,0 +1,68 @@
+        consoleWrite('PROMISE: licenseResponse resolved');
+        var license = await licenseResponse.text();
+
+        responseObject = JSON.parse(license.trim());
```
ditto

eric-carlson reviewed (commented):
I think it would be worth splitting this into two PRs: one with the new server and LayoutTests changes, and another with the WebKit code fixes/changes.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/index.py:
```diff
@@ -0,0 +1,146 @@
+import traceback
+
+body = sys.stdin.read()
+request = json.loads(body)
```
All the `json.loads` can be put within a try/catch block and you can just dump the body when an exception is thrown.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py:
```diff
@@ -0,0 +1,130 @@
+
+
+class Utils:
+    privateKey = None
```
There is name shadowing with this variable and it doesn't seem to be used anywhere as a class scoped attribute.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py:
```diff
@@ -0,0 +1,130 @@
+
+class Utils:
+    privateKey = None
+    ASK = None
```
This variable isn't used anywhere

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/index.py:
```diff
@@ -0,0 +1,146 @@
+
+    try:
+        uriComponents = urllib.parse.urlparse(uri)
+    except(e):
```
`except` in python takes in an exception type like `except ZeroDivisionError:` or blank `except:` to catch everything.

On top of this, once this exception is caught, the program will continue execution as normal, so I would probably return after catching an exception if we can't continue.

jernoble replied:
`badRequest()` will call `exit()`, so it won't continue after catching the exception.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/index.py:
```diff
@@ -0,0 +1,146 @@
+        ckcData = ckcContainer.serialize()
+
+        responseCKCs.append({"ckc": base64.b64encode(ckcData).decode('utf-8')})
+    except(e):
```
ditto.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/index.py:
```diff
@@ -0,0 +1,146 @@
+
+    assetId = uriComponents.netloc
+    if len(assetId) == 0:
+        badRequest("assetId '{}' is empty", assetId)
```
`"assetId '{}' is empty".format(assetId)`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py:
```diff
@@ -0,0 +1,130 @@
+        return privateKey
+
+    def GenRandom(size) -> bytearray:
+        return secrets.token_bytes(size)
```
`secrets.token_bytes` returns `bytes` not `bytearray`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py:
```diff
@@ -0,0 +1,130 @@
+        assert privateKey
+
+        cipher = PKCS1_OAEP.new(privateKey)
+        aesKey = cipher.decrypt(aesWrappedKey)
```
`Crypto.Cipher.PKCS1_OAEP.PKCS1OAEP_Cipher.decrypt` returns `bytes` not `bytearray`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py:
```diff
@@ -0,0 +1,130 @@
+import sys
+
+
+class Utils:
```
Use the `@staticmethod` decorator for all the methods in this class.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/Structs.py:
```diff
@@ -0,0 +1,47 @@
+class SKDAssetInfo:
+    assetId:          str = ""
+    contentKey:       array = bytearray(Constants.PS_AES128_KEY_SZ)
+    iv:               array = bytearray(Constants.PS_AES128_IV_SZ)
```
`bytearray =` instead of `array =`, different types.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/DFunction.py:
```diff
@@ -0,0 +1,93 @@
+
+
+def ComputeDFunctionHash(R2):
+    assert len(R2) > 0 and len(R2) < 56
```
`assert 0 < len(R2) < 56`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/DFunction.py:
```diff
@@ -0,0 +1,93 @@
+
+
+def DFunction(R2, ASk):
+    assert len(R2) > 0 and len(R2) < 56
```
ditto.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/SPCContainer.py:
```diff
@@ -0,0 +1,174 @@
+
+        return container
+
+    def __init__(self, spcVersion, spcIV, wrappedKey, certificateHash, payloadSize, payloadData):
```
`self._decryptedData` is missing

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLV.py:
```diff
@@ -0,0 +1,147 @@
+        Utils.SetBigEndian32(self.totalSize, serializedData, 8)
+        Utils.SetBigEndian32(self.valueSize, serializedData, 12)
+        serializedData[16:] = self._valueData
+        if (self._paddingData):
```
no need for parenthesis

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLV.py:
```diff
@@ -0,0 +1,147 @@
+
+    def serialize(self):
+        serializedData = bytearray(self.HEADER_SZ + self.totalSize)
+        offset = Offset(0)
```
`offset` is never used?

jernoble replied:
Removed.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @property
+    def isDecrypted(self):
+        return True if self._decryptedData else False
```
`self._decryptedData` isn't initialized in `__init__`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @value.setter
+    def value(self, value):
+        assert valueSize == 21
```
`self.valueSize`?

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @value.setter
+    def value(self, value):
+        assert valueSize == 16
```
ditto.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @value.setter
+    def value(self, value):
+        assert valueSize == 21
```
ditto.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+    TAG = 0x1bf7f53f5d5d5a1f
+
+    def parseValue(self):
+        assert self.valueSize >= 2 and self.valueSize <= 200
```
`assert 2 <= self.valueSize <= 200`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @value.setter
+    def value(self, value):
+        assert len(value) >= 2 and len(value) <= 200
```
`assert 2 <= len(value) <= 200`

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+    def value(self, value):
+        valueData = bytearray(8)
+        Utils.SetBigEndian64(value)
+        self.valueData = value
```
`valueData` is never used here.

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+
+    @value.setter
+    def value(self, value):
+        assert valueSize == 44
```
`self.valueSize`?

youssefsoli commented on LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py:
```diff
@@ -0,0 +1,494 @@
+        return ContentKeyDuration(self.TAG, totalSize, valueSize, valueData, padding)
+
+    def parseValue(self):
+        assert value >= 32
```
ditto.

eric-carlson reviewed (approved):
rs=me given Youssef's extensive review

jernoble committed:
[Cocoa] Add a FairPlay keyserver and EME tests to LayoutTests
https://bugs.webkit.org/show_bug.cgi?id=237731
<rdar://problem/90453705>

Reviewed by Eric Carlson.

LayoutTests:
Add an implementation of a FairPlay keyserver, which can run as a CGI script as part of
the normal run-webkit-httpd, to LayoutTests.

The Encrypted Media Extensions API has traditionally required access to an external
keyserver to respond to requests by the client for the encryption keys needed to
decrypt media content. However, our testing infrastructure requires that no external
network access is allowed (aside from localhost) from the process being tested.

A keyserver which can run locally will enable WebKit developers to write and run layout
tests exercising both the EME API and platform encrypted media playback capabilities,
and allows those tests to be run on our bots as both pre-submission checks and
as regression tests.

The initial keyserver implementation is written in python3, has minimal package requirements,
and currently only supports lease-style liceneses. It can be extended to support rental-style
and persistable licenses. It currently only supports "modern" EME key requests, but can
be extended to support "legacy" EME key requests as well.

The initial set of tests validate successful key exchange with 'skd', 'cenc', and 'sinf'
init data formats, key renewal, key rotation, multiple key support, and HLS playback.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-1.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-2.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-3.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-header-keyid-4.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-audio-payload.m4a: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-1.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-2.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-3.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-header-keyid-4.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/elementary-stream-video-payload.m4v: Added.
* LayoutTests/http/tests/media/fairplay/content/main.ts: Added.
* LayoutTests/http/tests/media/fairplay/content/prog_index.m3u8: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/prog_index.m3u8: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/segment-0.ts: Added.
* LayoutTests/http/tests/media/fairplay/content/rotation/segment-1.ts: Added.
* LayoutTests/http/tests/media/fairplay/eme2016.js: Added.
(async fetchBuffer):
(async getResponse):
(async startEME):
(async fetchAndAppend):
(async runAndWaitForLicenseRequest):
(async fetchAndWaitForLicenseRequest):
(async fetchAppendAndWaitForEncrypted):
(async createBufferAndAppend):
(async createBufferAppendAndWaitForEncrypted):
* LayoutTests/http/tests/media/fairplay/fps-hls-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-key-rotation-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-key-rotation.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-update-reject-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls-update-reject.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-hls.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-cenc-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-cenc.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-sinf-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-sinf.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-skd-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-init-data-skd.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-renewal-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-renewal.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-rotation-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-key-rotation.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-multiple-keys-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-multiple-keys.html: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-same-key-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-same-key.html: Added.
* LayoutTests/http/tests/media/fairplay/mse.js: Added.
(MSE):
(formatRangeHeader):
(MSE.prototype.start):
(MSE.prototype.handleSourceOpen):
(MSE.prototype.fetchNext):
(MSE.prototype.fetchChunk):
(MSE.prototype.handleChunkLoaded):
(MSE.prototype.handleChunkError):
(MSE.prototype.handleSourceBufferUpdateStart):
(MSE.prototype.handleSourceBufferUpdateEnd):
(MSE.prototype.removeAllSamples):
* LayoutTests/http/tests/media/fairplay/resources/GenerateCKC.py: Added.
* LayoutTests/http/tests/media/fairplay/resources/cert.der: Added.
* LayoutTests/http/tests/media/fairplay/resources/index.py: Added.
* LayoutTests/http/tests/media/fairplay/resources/keyserver/CKCContainer.py: Added.
(CKCContainer):
(CKCContainer.__init__):
(CKCContainer.version):
(CKCContainer.iv):
(CKCContainer.tllvs):
(CKCContainer.encryptedPayload):
(CKCContainer.addTLLV):
(CKCContainer.getTLLV):
(CKCContainer.encrypt):
(CKCContainer.serialize):
(CKCContainer.__str__):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Constants.py: Added.
(SKDServerAESEncType):
(SKDServerAESEncMode):
(SKDHDCPInformation):
(SKDClientRefTimePlayback):
(SKDLeaseRentalType):
(SKDStreamingIndicatorValue):
(SKDCapabilityFlags):
(Constants):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/DFunction.py: Added.
(ComputeDFunctionHash):
(Encrypt):
(DFunction):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/SPCContainer.py: Added.
(SPCContainer):
(SPCContainer.parse):
(SPCContainer.__init__):
(SPCContainer.spcVersion):
(SPCContainer.spcIV):
(SPCContainer.wrappedKey):
(SPCContainer.certificateHash):
(SPCContainer.payloadSize):
(SPCContainer.payloadData):
(SPCContainer.tllvs):
(SPCContainer.decrypt):
(SPCContainer.getTLLV):
(SPCContainer.checkIntegrity):
(SPCContainer.generateCKC):
(SPCContainer.generateARSeed):
(SPCContainer.__str__):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Structs.py: Added.
(Offset):
(SKDAssetInfo):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLV.py: Added.
(TLLV):
(TLLV.parse):
(TLLV.paddingForValueData):
(TLLV.__init__):
(TLLV.tagValue):
(TLLV.totalSize):
(TLLV.valueSize):
(TLLV.paddingSize):
(TLLV.valueData):
(TLLV.paddingData):
(TLLV.parseValue):
(TLLV.serialize):
(TLLV.__str__):
(Unknown):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/TLLVs.py: Added.
(SK_R1):
(SK_R1.parseValue):
(SK_R1.iv):
(SK_R1.payload):
(SK_R1.isDecrypted):
(SK_R1.sk):
(SK_R1.hu):
(SK_R1.r1):
(SK_R1.integrity):
(SK_R1.decrypt):
(SK_R1.encrypt):
(SK_R1.__str__):
(SK_R1_Integrity):
(SK_R1_Integrity.parseValue):
(SK_R1_Integrity.value):
(SK_R1_Integrity.__str__):
(Capbilities):
(Capbilities.parseValue):
(Capbilities.capabilityFlags):
(Capbilities.has):
(Capbilities.set):
(Capbilities.clear):
(AntiReplaySeed):
(AntiReplaySeed.parseValue):
(AntiReplaySeed.value):
(R2):
(R2.parseValue):
(R2.value):
(ReturnTagRequest):
(ReturnTagRequest.parseValue):
(ReturnTagRequest.returnTags):
(AssetID):
(AssetID.parseValue):
(AssetID.value):
(TransactionID):
(TransactionID.parseValue):
(ProtocolVersionsSupported):
(ProtocolVersionsSupported.parseValue):
(ProtocolVersionsSupported.supportedVersions):
(ProtocolVersionUsed):
(ProtocolVersionUsed.parseValue):
(ProtocolVersionUsed.usedVersion):
(StreamingIndicator):
(StreamingIndicator.parseValue):
(StreamingIndicator.value):
(MediaPlaybackState):
(MediaPlaybackState.parseValue):
(MediaPlaybackState.creationDate):
(MediaPlaybackState.playbackState):
(MediaPlaybackState.sessionID):
(EncryptedContentKey):
(EncryptedContentKey.create):
(EncryptedContentKey.parseValue):
(EncryptedContentKey.iv):
(EncryptedContentKey.ck):
(EncryptedContentKey.__str__):
(R1):
(R1.create):
(R1.parseValue):
(R1.value):
(ContentKeyDuration):
(ContentKeyDuration.create):
(ContentKeyDuration.parseValue):
(ContentKeyDuration.leaseDuration):
(ContentKeyDuration.rentalDuration):
(ContentKeyDuration.keyType):
(ContentKeyDuration.__str__):
(HDCPEnforcement):
(HDCPEnforcement.parseValue):
(HDCPEnforcement.value):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/Utils.py: Added.
(Utils):
(Utils.LoadPrivateKey):
(Utils.GenRandom):
(Utils.RSADecryptKey):
(Utils.AESEncryptDecrypt):
(Utils.FetchContentKeyAndIV):
(Utils.GetSupportedVersions):
(Utils.GetBigEndian32):
(Utils.GetBigEndian64):
(Utils.SetBigEndian32):
(Utils.SetBigEndian64):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/__init__.py: Added.
(cd_dot_dot):
(_maybe_add_scripts_path):
(_force_cryptography_install):
(_force_openssl_install):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_DFunction.py: Added.
(TestDFunction):
(TestDFunction.test_DFunctionHash):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_TLLV.py: Added.
(TestTLLV):
(TestTLLV.test_TLLV):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_TLLVs.py: Added.
(TestTLLVs):
(TestTLLVs.test_SK_R1):
(TestTLLVs.test_EncryptedContentKey):
(TestTLLVs.test_Capabilities):
(TestTLLVs.test_ReturnTagRequest):
(TestTLLVs.test_AssetID):
* LayoutTests/http/tests/media/fairplay/resources/keyserver/test_Utils.py: Added.
(TestUtils):
* LayoutTests/http/tests/media/fairplay/support.js: Added.
(uInt8ArrayToString):
(base64DecodeUint8Array):
(stringToUInt8Array):
(base64EncodeUint8Array):
(failTestWithException):
* LayoutTests/platform/ios-simulator/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:

Canonical link: https://commits.webkit.org/253230@main
